### PR TITLE
Account for trait alias when looking for defid

### PR DESCRIPTION
**Pull request by lzutao**
I hit the crash on the `expect` call when running clippy on rustc libcore.
Hopefully this will fix it.
changelog: none

### CODE REVIEW

**Comment by phansch** [comment]
Do you think it would be a lot of work to extract a minimal test case? It would be nice to have a test for this

**Comment by lzutao** [comment]
Do clippy have a way to log the file it is processing? I think this would help extract a testcase.
Or just merge this and test again on libcore I think.

**Comment by mikerite** [comment]
Just apply the changes locally and test with your local copy.

```
cd "<project_dir>"
cargo build --manifest-path "<local_clippy>/Cargo.toml"
cargo run --manifest-path "<local_clippy>/Cargo.toml" --bin cargo-clippy
```

**Comment by lzutao** [comment]
Thanks @mikerite. I cannot manage to find a testcase for this but binary
buillt with this patch fixes the crash when running on libcore.

**Comment by mikerite** [comment]
I found the bug. It's in `SelfKind::matches_ref ` where we look up the `AsRef` or `AsMut` traits. libcore is probably one of the few crates where these don't exist. It wouldn't be easy to make a test case for this. You'd have to build something without using `core`.

A more targetted fix would be:

```
--- a/clippy_lints/src/methods/mod.rs
+++ b/clippy_lints/src/methods/mod.rs
@@ -2781,8 +2781,11 @@ impl SelfKind {
                 hir::Mutability::MutMutable => &paths::ASMUT_TRAIT,
             };
 
-            let trait_def_id = get_trait_def_id(cx, trait_path).expect("trait def id not found");
-            implements_trait(cx, ty, trait_def_id, &[parent_ty.into()])
+            if let Some(trait_def_id) = get_trait_def_id(cx, trait_path) {
+                implements_trait(cx, ty, trait_def_id, &[parent_ty.into()])
+            } else {
+                false
+            }
         }

```

**Comment by flip1995** [comment]
`#![no_core]` isn't a thing yet, is it?

**Comment by flip1995** [comment]
The example code, that I posted, already passes the test, without your change. You have to adapt this code, so that it matches
https://github.com/rust-lang/rust-clippy/blob/b8e5e6f3c204877da9c7eaef0d956514f2c8f6d2/clippy_lints/src/methods/mod.rs#L1074-L1086

https://github.com/rust-lang/rust-clippy/blob/b8e5e6f3c204877da9c7eaef0d956514f2c8f6d2/clippy_lints/src/methods/mod.rs#L1103-L1107

Where the last line in the second snippet should fail at the `matches` call

**Comment by flip1995** [comment]
I added it myself. Thanks for your fix!

@bors r+

File: tests/ui/def_id_nocore.stderr [hunk 1]
```diff
@@ -0,0 +1,10 @@
+error: methods called `as_*` usually take self by reference or self by mutable reference; consider choosing a less ambiguous name
+  --> $DIR/def_id_nocore.rs:26:19
+   |
+LL |     pub fn as_ref(self) -> &'static str {
+   |                   ^^^^
+   |
+   = note: `-D clippy::wrong-self-convention` implied by `-D warnings`
```
Comment (flip1995) [inline tests/ui/def_id_nocore.stderr]: Fun fact: If you add the `&` to `self` in a `no_core` crate, rustc complains that 
```
error[E0658]: `&A` cannot be used as the type of `self` without the `arbitrary_self_types` feature
  --> tests/ui/def_id_nocore.rs:26:19
   |
26 |     pub fn as_ref(&self) -> &'static str {
   |                   ^^^^^
   |
   = note: for more information, see https://github.com/rust-lang/rust/issues/44874
   = help: add `#![feature(arbitrary_self_types)]` to the crate attributes to enable
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)
```

I don't think we need to adapt the warning though :smile:

**Pull request merged**